crates/rethnet_evm/src/blockchain/local.rs

```diff
@@ -0,0 +1,219 @@
+    ///
+    /// # Safety
+    ///
+    /// Ensure that the genesis block's number is zero.
+    pub unsafe fn with_genesis_block_unchecked(genesis_block: DetailedBlock) -> Self {
+        let total_difficulty = genesis_block.header.difficulty;
+        let storage = ContiguousBlockchainStorage::with_block(genesis_block, total_difficulty);
+
+        Self { storage }
+    }
+}
+
+impl Blockchain for LocalBlockchain {
+    type Error = BlockchainError;
+
+    fn block_by_hash(&self, hash: &B256) -> Result<Option<Arc<DetailedBlock>>, Self::Error> {
+        Ok(self.storage.block_by_hash(hash).cloned())
+    }
+
+    fn block_by_number(&self, number: &U256) -> Result<Option<Arc<DetailedBlock>>, Self::Error> {
+        let number = usize::try_from(number).map_err(|_| BlockchainError::BlockNumberTooLarge)?;
+
+        Ok(self.storage.blocks().get(number).cloned())
+    }
+
+    fn block_by_transaction_hash(
+        &self,
+        transaction_hash: &B256,
+    ) -> Result<Option<Arc<DetailedBlock>>, Self::Error> {
+        Ok(self
+            .storage
+            .block_by_transaction_hash(transaction_hash)
+            .cloned())
+    }
+
+    fn last_block(&self) -> Result<Arc<DetailedBlock>, Self::Error> {
+        Ok(self
+            .storage
+            .blocks()
+            .last()
+            .expect("A genesis block is always present")
+            .clone())
+    }
+
+    fn last_block_number(&self) -> U256 {
+        // The block number of the genesis block is 0, so subtract one
+        U256::from(self.storage.blocks().len() - 1)
+    }
+
+    fn total_difficulty_by_hash(&self, hash: &B256) -> Result<Option<U256>, Self::Error> {
+        Ok(self.storage.total_difficulty_by_hash(hash).cloned())
+    }
+}
+
+impl BlockchainMut for LocalBlockchain {
+    type Error = BlockchainError;
+
+    fn insert_block(&mut self, block: DetailedBlock) -> Result<Arc<DetailedBlock>, Self::Error> {
+        let last_block = self.last_block()?;
+
+        let next_block_number = last_block.header.number + U256::from(1);
+        if block.header.number != next_block_number {
+            return Err(BlockchainError::InvalidBlockNumber {
+                actual: block.header.number,
+                expected: next_block_number,
+            });
+        }
+
+        if block.header.parent_hash != last_block.header.hash() {
+            return Err(BlockchainError::InvalidParentHash);
+        }
+
+        let previous_total_difficulty = self
+            .storage
+            .total_difficulties()
+            .last()
+            .expect("Storage always contains at least one block");
+
+        let total_difficulty = previous_total_difficulty + block.header.difficulty;
+
+        // SAFETY: The block number is guaranteed to be unique, so the block hash must be too.
+        let block = unsafe { self.storage.insert_block_unchecked(block, total_difficulty) };
+
+        Ok(block.clone())
+    }
```

Comment:
If you look at the two implementations, you'll notice that there are actually slight differences based on the underlying storage containers, that cannot be de-duplicated. In particular, search based on hash vs search based on index (sparse vs contiguous storage).

I did move the block validation to a shared function. To do so, I added caching of the hash, an optimisation that was needed anyway, to prevent the block hash from being calculated multiple times per call: https://github.com/NomicFoundation/hardhat/pull/4175/commits/a31a065963c6c1a2dc2b31e405829f6608ab59d8